src/mono/mono/metadata/class.c

```diff
@@ -1980,16 +1984,45 @@ mono_class_interface_offset_with_variance (MonoClass *klass, MonoClass *itf, gbo
 			return -1;
 
 		return m_class_get_interface_offsets_packed (klass) [found];
-	}
+	} else if (has_variance) {
+		MonoClass **vst;
+		int vst_count;
+		MonoClass *current = klass;
+
+		// Perform two passes per class, then check the base class
+		while (current) {
+			mono_class_get_variance_search_table (current, &vst, &vst_count);
+
+			// Exact match pass: Is there an exact match at this level of the type hierarchy?
+			// If so, we can use the interface_offset we computed earlier, since we're walking from most derived to least.
+			for (i = 0; i < vst_count; i++) {
+				if (itf != vst [i])
+					continue;
+
+				*non_exact_match = FALSE;
+				return exact_match;
+			}
 
-	if (!mono_class_has_variant_generic_params (itf))
-		return -1;
+			// Inexact match (variance) pass:
+			// Is any interface at this level of the type hierarchy variantly compatible with the desired interface?
+			// If so, select the first compatible one we find.
+			for (i = 0; i < vst_count; i++) {
+				if (!mono_class_is_variant_compatible (itf, vst [i], FALSE))
```

Comment:
Question: how is this variance search table different from the `interfaces_packed` table?  Do we really need both?  if the variance search table is more correct, can we get rid of the interfaces_packed table?

Reply:
The variance search table should be smaller. I can try to figure out whether we can get rid of interfaces_packed

Reply:
Two examples of what I was thinking in action:
```
  System.IBinaryFloatParseAndFormatInfo`1 vst_size=1 io_count=39
  System.Single vst_size=1 io_count=40
```